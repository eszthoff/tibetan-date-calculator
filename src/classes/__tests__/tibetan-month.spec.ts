import MockDate from 'mockdate';
import TibetanMonth from '../tibetan-month';

describe('TibetanMonth', () => {
  beforeAll(() => {
    MockDate.set('2019/06/07');
  });
  afterAll(() => {
    MockDate.reset();
  });

  it('should create a correct class when invoked without arg', () => {
    expect(new TibetanMonth()).toMatchSnapshot();
  });
  it('should create a correct class when invoked with string arg', () => {
    expect(new TibetanMonth('2019/06/22')).toMatchSnapshot();
  });
  it('should create a correct class when invoked with object arg', () => {
    expect(new TibetanMonth({
      year: 2146, month: 3, isLeapMonth: false
    })).toMatchSnapshot();
  });
  it('should set isLeapMonth to true for first of doubled months', () => {
    const month = new TibetanMonth('2019/02/10');
    expect(month.isDoubledMonth).toBeTruthy();
    expect(month.isLeapMonth).toBeTruthy();
  });
  it('should set isLeapMonth to true for second of doubled months', () => {
    const month = new TibetanMonth('2019/03/10');
    expect(month.isDoubledMonth).toBeTruthy();
    expect(month.isLeapMonth).toBeFalsy();
  });
  it('should not calculate days when initially incoved', () => {
    const month = new TibetanMonth();
    expect(month.days.length).toEqual(0);
  });
  it('should calculate days when requested (includes skipped and doubled days)', () => {
    const month = new TibetanMonth({ year: 2146, month: 5 });
    const days = month.getDays();

    expect(month.days.length).not.toEqual(0);
    expect(days).toMatchSnapshot();
  });
  it('should not include skipped days', () => {
    const month = new TibetanMonth({ year: 2146, month: 5 });
    const days = month.getDays();

    expect(days.some(day => day.isPreviousSkipped)).toBeTruthy();
    expect(days.every(day => !day.isSkippedDay)).toBeTruthy();
  });
  it('should list days in the correct order according to tibetan date', () => {
    const month = new TibetanMonth({ year: 2146, month: 5 });
    const days = month.getDays();
    const isDateAfterPrev = (day, i, arr): boolean => {
      if (i === 0 && day.date === 1) {
        return true;
      }

      return day.date >= arr[i - 1].date;
    };

    expect(days.every(isDateAfterPrev)).toBeTruthy();
  });
  it('should list days in the correct order according to western date', () => {
    const month = new TibetanMonth({ year: 2146, month: 5 });
    const days = month.getDays();
    const isDateAfterPrev = (day, i, arr): boolean => {
      if (i === 0 && day.date === 1) {
        return true;
      }

      return day.westernDate > arr[i - 1].westernDate;
    };

    expect(days.every(isDateAfterPrev)).toBeTruthy();
  });
  it('should return year object correctly)', () => {
    const month = new TibetanMonth();
    const { yearObj } = month;

    expect(yearObj).toMatchSnapshot();
  });
  it('should return the correct string when `toString` is called in not double month', () => {
    expect(new TibetanMonth().toString()).toEqual('2146-4');
  });
  it('should return the correct string when `toString` is called in leap of double month', () => {
    expect(new TibetanMonth('2019/02/10').toString()).toEqual('2146-1-leap');
  });
  it('should return the correct string when `toString` is called in main of double month', () => {
    expect(new TibetanMonth('2019/03/10').toString()).toEqual('2146-1-main');
  });
});
