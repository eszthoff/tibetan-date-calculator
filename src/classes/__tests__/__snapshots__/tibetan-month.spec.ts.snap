// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TibetanMonth should calculate days when requested (includes skipped and doubled days) 1`] = `
Array [
  TibetanDate {
    "date": 1,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-03T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 2,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-04T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 3,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-05T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 4,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-06T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 5,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-07T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 6,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-08T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 7,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-09T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 8,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-10T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 10,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": true,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-11T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 11,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-12T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 12,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-13T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 13,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-14T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 14,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-15T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 15,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-16T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 16,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-17T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 17,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-18T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 18,
    "isDoubledDay": true,
    "isLeapDay": true,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-19T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 18,
    "isDoubledDay": true,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-20T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 19,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-21T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 20,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-22T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 21,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-23T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 22,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-24T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 23,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-25T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 24,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-26T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 25,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-27T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 26,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-28T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 27,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-29T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 28,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-30T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 29,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-07-31T10:00:00.000Z,
    "year": 2146,
  },
  TibetanDate {
    "date": 30,
    "isDoubledDay": false,
    "isLeapDay": false,
    "isPreviousSkipped": false,
    "isSkippedDay": false,
    "month": 5,
    "monthObj": TibetanMonth {
      "days": Array [],
      "endDateStr": "2019-08-01",
      "isDoubledMonth": false,
      "isLeapMonth": false,
      "month": 5,
      "startDateStr": "2019-07-03",
      "year": 2146,
    },
    "westernDate": 2019-08-01T10:00:00.000Z,
    "year": 2146,
  },
]
`;

exports[`TibetanMonth should create a correct class when invoked with object arg 1`] = `
TibetanMonth {
  "days": Array [],
  "endDateStr": "2019-06-03",
  "isDoubledMonth": false,
  "isLeapMonth": false,
  "month": 3,
  "startDateStr": "2019-05-05",
  "year": 2146,
}
`;

exports[`TibetanMonth should create a correct class when invoked with string arg 1`] = `
TibetanMonth {
  "days": Array [],
  "endDateStr": "2019-07-02",
  "isDoubledMonth": false,
  "isLeapMonth": false,
  "month": 4,
  "startDateStr": "2019-06-04",
  "year": 2146,
}
`;

exports[`TibetanMonth should create a correct class when invoked without arg 1`] = `
TibetanMonth {
  "days": Array [],
  "endDateStr": "2019-07-02",
  "isDoubledMonth": false,
  "isLeapMonth": false,
  "month": 4,
  "startDateStr": "2019-06-04",
  "year": 2146,
}
`;

exports[`TibetanMonth should return year object correctly) 1`] = `
TibetanYear {
  "animal": "Pig",
  "element": "Earth",
  "gender": "Female",
  "months": Array [],
  "rabjungCycle": 17,
  "rabjungYear": 33,
  "tibYearNum": 2146,
  "westernYear": 2019,
}
`;
